sketchlib/AnimationChain.js

```diff
@@ -0,0 +1,188 @@
+    // Get lines oriented away from the center point.
+    // Note that these are normalized
+    const ba = b.join(a);
+    const bc = b.join(c);
+
+    // If the bend angle is mostly straight, just do the simpler following
+    // behavior
+    const dot_product = bc.dot(ba);
+    const max_dot_product = Math.cos(min_bend_angle);
+    if (dot_product < max_dot_product) {
+      return this.constraint_follow(b, c, follow_distance);
+    }
+
+    const is_ccw = bc.sin_angle_to(ba) > 0;
+    const unbent_angle = is_ccw ? -min_bend_angle : min_bend_angle;
+    const forward = a.sub(b).normalize();
+    const rotation = Motor.rotation(Point.ORIGIN, unbent_angle);
+    const offset = rotation.transform_point(forward).scale(follow_distance);
+    return b.add(offset);
+  }
+}
+
+/**
+ * A chain of joints. When you move the head of the chain, the rest of it
+ * follows behind it.
+ *
+ * based on https://www.youtube.com/watch?v=qlfh_rv6khY&ab_channel=argonaut
+ * but implemented using PGA 2D
+ */
+export class AnimationChain {
+  /**
+   * Constructor
+   * @param {Joint[]} joints The joints in the chain. The first joint is the head.
+   * @param {number} min_bend_angle The minimum allowed bending angle so
+   * the curve doesn't fold up and self-intersect
+   */
+  constructor(joints, min_bend_angle) {
+    this.joints = joints;
+    this.min_bend_angle = min_bend_angle;
+  }
+
+  get head() {
+    return this.joints[0];
+  }
+
+  get tail() {
+    return this.joints[this.joints.length - 1];
+  }
+
+  /**
+   * Get the number of joints in the chain
+   */
+  get length() {
+    return this.joints.length;
+  }
+
+  /**
+   * Get a specific joint
+   * @param {number} i The joint index
+   */
+  get_joint(i) {
+    if (i < 0) {
+      return this.joints[this.joints.length - 1 + i];
+    }
+    return this.joints[i];
+  }
+
+  /**
+   * Get an array of the joint positions
+   * @returns {Point[]} The positions of the joints
+   */
+  get_positions() {
+    return this.joints.map((x) => x.position);
+  }
+
+  render_spine() {
+    const lines = new Array(this.joints.length - 1);
+    const positions = this.get_positions();
+    for (let i = 0; i < lines.length; i++) {
+      lines[i] = new LinePrimitive(positions[i], positions[i + 1]);
+    }
+    const spine_group = new GroupPrimitive(lines, SPINE_STYLE);
+
+    const centers = positions.map((x) => new PointPrimitive(x));
+    const centers_group = new GroupPrimitive(centers, CENTER_STYLE);
+
+    return new GroupPrimitive([spine_group, centers_group]);
+  }
+
+  /**
+   * Move the head of the chain to target, then move the rest of the chain
+   * behind it.
+   * @param {Point} target The new position for the head joint
+   */
+  move(target) {
+    // If we're already at the target, nothing to do!
+    if (!this.head.position.equals(target)) {
+      this.head.follow(target);
+    }
+
+    for (let i = 1; i < this.joints.length; i++) {
```

Comment:
Add a comment explaining that for cases where the chain is changing size, we still want to update the rest of the chain